### PR TITLE
Review feed (accessible by everyone)

### PR DESCRIPTION
**Pull request by archthegit**
* I still have rating to refine, right now its just a text box , want to change it to stars
* I changed the type for rating attribute in review from Integer to Long because it was throwing me an error when I called createListOfReviews() method. I couldn't find the issue, if someone can help that'd be great
* made a temp book page which works the same exact way as user-page but just posts review objects instead of message objects. This can later be used when you need a page to post reviews after user clicks on book title.

### CODE REVIEW

File: src/main/java/com/google/codeu/data/Book.java [hunk 1]
```diff
@@ -69,14 +69,15 @@ public double getAvgRating() {
...
    */
   public double calcAvgRating(List<Review> reviews) {
+    if (reviews.size() == 0) {
+      return 0;
+    }
     double average = 0.0;
-    for (int i = 0; i < reviews.size(); i++) {
-      average += reviews.get(i).getRating();
+    for (Review review: reviews) {
+      average += review.getRating();
```
Comment (karenfrilya97) [inline src/main/java/com/google/codeu/data/Book.java]: I think you should call this variable `sum` instead of `average` because it is really a sum of the reviews' ratings, and calling it `average` can be confusing to people who read your code later on.

File: src/main/java/com/google/codeu/servlets/ReviewFeedServlet.java [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+import java.io.IOException;
+import java.util.List;
+
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+
+@WebServlet("/reviewFeed")
```
Comment (karenfrilya97) [inline src/main/java/com/google/codeu/servlets/ReviewFeedServlet.java]: Since the rest of the servlets use lowercase with hyphen, it might be better to change this to `review-feed` instead to standardise.

File: src/main/java/com/google/codeu/servlets/ReviewServlet.java [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+public class ReviewServlet extends HttpServlet {
+  private Datastore datastore;
+
+  @Override
+  public void init() {
+    datastore = new Datastore();
+  }
+
+  /**
+   * Responds with a JSON representation of {@link Message} data for a specific user. Responds with
```
Comment (karenfrilya97) [inline src/main/java/com/google/codeu/servlets/ReviewServlet.java]: Do you mean `{@link Review}` instead of `{@link Message}`?

File: src/main/webapp/book-page.html [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+    <div class="column column-20">
+      <img id="profile-pic" class="avatar" alt="Profile Picture">
+      <div id="profile-pic-editor" class="edit-btn text-center hidden" onclick="
+                  show('profile-pic-form');
+                  hide('profile-pic-editor');
+                ">
+        Edit
+      </div>
+      <form id="profile-pic-form" action="/profile-pic" method="POST" enctype="multipart/form-data" class="hidden">
+        Upload new profile picture:
```
Comment (karenfrilya97) [inline src/main/webapp/book-page.html]: I think we should change Profile Picture here to Book Picture or something because profile picture is linked to a User and that won't be right for a book page.

I understand you made this book-page.html as a temporary page to test your Review Feed codes. But since you already did it, maybe we can just use it and then we might as well fix the Profile Pic here.

Comment (karenfrilya97) [reply src/main/webapp/book-page.html]: But you can ignore this comment if this file is gonna be totally replaced by another book-page.html

File: src/main/webapp/js/book-page-loader.js [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+        show('review-form');
+        show('profile-pic-editor');
+        show('about-me-editor');
+      }
+    });
+}
+
+/** Fetches messages and adds them to the page. */
+function fetchReviews() {
+  const url = `/reviews?user=${parameterUsername}`;
```
Comment (karenfrilya97) [inline src/main/webapp/js/book-page-loader.js]: I'm slightly confused about this. I thought book page should show the details of a book. So the reviews should be reviews for a `Book` not reviews for a `User`?

Maybe we can include the reviews of a `User` as part of the `user-page` instead. Maybe have 2 separate tabs, one for messages and one for reviews made by the user.

File: src/main/java/com/google/codeu/data/Book.java (hunk 1 above)
Comment (archthegit) [reply src/main/java/com/google/codeu/data/Book.java]: Yeap, I agree.

File: src/main/java/com/google/codeu/servlets/ReviewServlet.java (hunk 3 above)
Comment (archthegit) [reply src/main/java/com/google/codeu/servlets/ReviewServlet.java]: Lol, yeah will change that

File: src/main/webapp/book-page.html (hunk 4 above)
Comment (archthegit) [reply src/main/webapp/book-page.html]: No I was planning to turn this into a book-page template anyway, i'll fix this

File: src/main/webapp/js/book-page-loader.js (hunk 5 above)
Comment (archthegit) [reply src/main/webapp/js/book-page-loader.js]: Yeah thing is, at the time of writing this code, there wasn't any logic to put books into a data store. I'll fix this

File: src/main/java/com/google/codeu/data/Datastore.java [hunk 6]
```diff
@@ -131,6 +171,33 @@ public User getUser(String email) {
...
+
+    return createListOfReviews(results, null);
+  }
+
+  /**
+   * Gets reviews posted by a specific user.
+   *
+   * @return a list of reviews posted by the user, or empty list if user has never posted a review.
+   */
+  public List<Review> getReviews(String user) {
```
Comment (SuyashLakhotia) [inline src/main/java/com/google/codeu/data/Datastore.java]: `getReviewsForUser()`

File: src/main/java/com/google/codeu/data/Datastore.java [hunk 7]
```diff
@@ -130,8 +117,11 @@ public void storeReview(Review review) {
         long rating = (Long) entity.getProperty("rating");
```
Comment (karenfrilya97) [inline src/main/java/com/google/codeu/data/Datastore.java]: I just realised that `createListOfReviews` and `createListOfMessages` should be private because they are helper methods that are only used in this file.

File: src/main/java/com/google/codeu/data/Datastore.java [hunk 8]
```diff
@@ -48,20 +48,7 @@ public void storeMessage(Message message) {
   }
 
   /**
-   * Stores the Review in Datastore.
-   */
-  public void storeReview(Review review) {
```
Comment (karenfrilya97) [inline src/main/java/com/google/codeu/data/Datastore.java]: Why did you delete this method? Or did you move it somewhere else?

File: src/main/java/com/google/codeu/data/Datastore.java [hunk 9]
```diff
@@ -72,9 +72,67 @@ public void storeMessage(Message message) {
...
+
+    return messages;
+  }
+
+  /**
+   * Creates a list of reviews with or without id.
+   *
+   * @return a list of all reviews posted or by a specific user
+   */
+  public List<Review> createListOfReviews(PreparedQuery results, String user) {
```
Comment (karenfrilya97) [inline src/main/java/com/google/codeu/data/Datastore.java]: I just realised that `createListOfReviews` and `createListOfMessages` should be private because they are helper methods that are only used in this file.

File: src/main/java/com/google/codeu/data/Review.java [hunk 10]
```diff
@@ -72,10 +72,12 @@ public void setAuthor(String author) {
   }
 
   public long getRating() {
+
```
Comment (karenfrilya97) [inline src/main/java/com/google/codeu/data/Review.java]: Why did you add all these lines? Is it a coding guideline or something?

File: src/main/java/com/google/codeu/servlets/ReviewServlet.java [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+   * Responds with a JSON representation of {@link Review} data for a specific user. Responds with
+   * an empty array if the user is not provided.
+   */
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    response.setContentType("application/json");
+    List<Review> reviews = datastore.getAllReviews();
+    Gson gson = new Gson();
+    String json = gson.toJson(reviews);
+    response.getWriter().println(json);
```
Comment (karenfrilya97) [inline src/main/java/com/google/codeu/servlets/ReviewServlet.java]: Is there a reason why you use `getWriter` here instead of `getOutputStream` like in the other servlets? Just curious.

File: src/main/java/com/google/codeu/servlets/ReviewServlet.java [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    UserService userService = UserServiceFactory.getUserService();
+    if (!userService.isUserLoggedIn()) {
+      response.sendRedirect("/index.html");
+      return;
+    }
+
+    String user = userService.getCurrentUser().getEmail();
+    String text = Jsoup.clean(request.getParameter("text"), Whitelist.none());
+    long rating = Long.parseLong(Jsoup.clean(request.getParameter("rating"), Whitelist.none()));
```
Comment (karenfrilya97) [inline src/main/java/com/google/codeu/servlets/ReviewServlet.java]: Is it necessary/useful to use `Jsoup.clean` here if you are calling `parseLong`?

File: src/main/webapp/index.html [hunk 13]
```diff
@@ -36,9 +36,9 @@
       <li><a href="/">Home</a></li>
       <li><a href="/aboutus.html">About Our Team</a></li>
       <li><a href="/feed.html">Public Feed</a></li>
+      <li><a href="/review-feed.html">ReviewFeed</a> </li>
```
Comment (karenfrilya97) [inline src/main/webapp/index.html]: I think it's missing a space between Review and Feed.

**Comment by jannatbedi** [comment]
Just a suggestion..is it possible ton use books instead of stars for rating?

**Comment by archthegit** [comment]
> Just a suggestion..is it possible to use books instead of stars for rating?

lol sure, neither one has been implemented yet, if you'd like to give me a hand?

**Comment by karenfrilya97** [comment]
I'm doing the rating UI for the book list! I can add it here later, too. And, I'm currently using stars but can change to books later.

**Comment by karenfrilya97** [comment]
I'm doing the rating UI for the book list! I can add it here later, too. And, I'm currently using stars but can change to books later.

File: src/main/java/com/google/codeu/data/Review.java (hunk 10 above)
Comment (archthegit) [reply src/main/java/com/google/codeu/data/Review.java]: Happened automatically lol by intellij, I'll get it out

**Comment by archthegit** [comment]
> I'm doing the rating UI for the book list! I can add it here later, too. And, I'm currently using stars but can change to books later.

@karenfrilya97 let me know if i can help with anything!

File: src/main/java/com/google/codeu/servlets/ReviewServlet.java (hunk 11 above)
Comment (archthegit) [reply src/main/java/com/google/codeu/servlets/ReviewServlet.java]: No reason really, its what i found in the guided project for public feed. It's used in messageServlet's doGet method

**Comment by jannatbedi** [comment]
> > Just a suggestion..is it possible to use books instead of stars for rating?
> 
> @jannatbedi lol sure, neither one has been implemented yet, if you'd like to give me a hand?

@archthegit Sure! let me know how I can help you!

**Comment by jannatbedi** [comment]
> I'm doing the rating UI for the book list! I can add it here later, too. And, I'm currently using stars but can change to books later.

That would be great! Thanks @karenfrilya97 ! Let me know if you need any help!

File: src/main/java/com/google/codeu/servlets/ReviewServlet.java (hunk 11 above)
Comment (karenfrilya97) [reply src/main/java/com/google/codeu/servlets/ReviewServlet.java]: Hmm interesting. The other guided projects use `getOutputStream()`.But I guess there is not much diff either way.
